ecosystem/typescript/sdk/examples/typescript-esm/offer_capabilities.ts

```diff
@@ -0,0 +1,134 @@
+    const { data } = await provider.aptosClient.getAccountResource(alice.address(), "0x1::account::Account");
+    const offerFor = (data as any).rotation_capability_offer.for.vec[0];
+    console.log({ rotation_capability_offer: { for: offerFor }, bob: bob.address().toString() });
+    assert(offerFor.toString() == bob.address().toString(), "Bob's address should be in the rotation capability offer");
+  }
+
+  // Offer signer capability
+  {
+    // Create and fund two new accounts
+    const { alice, bob } = await createAndFundAliceAndBob(faucetClient);
+    // Note that the signer capability offer doesn't require the chainId
+    console.log(`---------------  SignerCapabilityOfferProofChallengeV2 ---------------`);
+    const { hash, version, success, payload } = await signStructAndSubmitTransaction(
+      provider,
+      alice,
+      "offer_signer_capability",
+      {
+        accountAddress: TxnBuilderTypes.AccountAddress.fromHex(CORE_CODE_ADDRESS),
+        moduleName: "account",
+        structName: "SignerCapabilityOfferProofChallengeV2",
+        sequenceNumber: Number((await provider.aptosClient.getAccount(alice.address())).sequence_number),
+        sourceAddress: TxnBuilderTypes.AccountAddress.fromHex(alice.address()),
+        recipientAddress: TxnBuilderTypes.AccountAddress.fromHex(bob.address()),
+      },
+    );
+    console.log({ hash, version, success, payload });
+    const { data } = await provider.aptosClient.getAccountResource(alice.address(), "0x1::account::Account");
+    const offerFor = (data as any).signer_capability_offer.for.vec[0];
+    console.log({ signer_capability_offer: { for: offerFor }, bob: bob.address().toString() });
+    assert(offerFor.toString() == bob.address().toString(), "Bob's address should be in the signer capability offer");
+  }
+})();
+
+const signStructAndSubmitTransaction = async (
+  provider: Provider,
+  signer: AptosAccount,
+  funcName: string,
+  struct: CapabilityOfferProofChallengeV2,
+): Promise<any> => {
+  // The proof bytes are just the individual BCS serialized
+  // data concatenated into a single byte array.
+  // Note that the proof bytes must be constructed in this specific order- the order of the struct data on-chain.
+  const proofBytes = new Uint8Array([
+    ...BCS.bcsToBytes(struct.accountAddress),
+    ...BCS.bcsSerializeStr(struct.moduleName),
+    ...BCS.bcsSerializeStr(struct.structName),
+    ...(struct.chainId ? BCS.bcsSerializeU8(struct.chainId) : []),
+    ...BCS.bcsSerializeUint64(struct.sequenceNumber),
+    ...BCS.bcsToBytes(struct.sourceAddress),
+    ...BCS.bcsToBytes(struct.recipientAddress),
+  ]);
+
+  // This is the actual signature of the struct.
+  const signedMessage = signer.signBuffer(proofBytes).toUint8Array();
+
+  // Note the hard-coded account scheme. You may need to accomodate MultiEd25519.
+  const payload = new TxnBuilderTypes.TransactionPayloadEntryFunction(
+    TxnBuilderTypes.EntryFunction.natural(
```

Comment:
Never to this day will I understand what natural means in this context.

Reply:
🤣